src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp

```diff
@@ -110,168 +155,302 @@ class boss_thaddius : public CreatureScript
         return GetInstanceAI<boss_thaddiusAI>(creature);
     }
 
-    struct boss_thaddiusAI : public BossAI
+    class boss_thaddiusAI : public BossAI
     {
-        boss_thaddiusAI(Creature* creature) : BossAI(creature, BOSS_THADDIUS)
-        {
-            // init is a bit tricky because thaddius shall track the life of both adds, but not if there was a wipe
-            // and, in particular, if there was a crash after both adds were killed (should not respawn)
-
-            // Moreover, the adds may not yet be spawn. So just track down the status if mob is spawn
-            // and each mob will send its status at reset (meaning that it is alive)
-            checkFeugenAlive = false;
-            if (Creature* pFeugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
-                checkFeugenAlive = pFeugen->IsAlive();
-
-            checkStalaggAlive = false;
-            if (Creature* pStalagg = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_STALAGG)))
-                checkStalaggAlive = pStalagg->IsAlive();
+        public:
+            boss_thaddiusAI(Creature* creature) : BossAI(creature, BOSS_THADDIUS), _phase(PHASE_NOT_ENGAGED), stalaggAlive(true), feugenAlive(true), ballLightningEnabled(false), shockingEligibility(true) {
+                if (instance->GetBossState(BOSS_THADDIUS) != DONE)
+                {
+                    if (!me->IsAlive()) // this can happen if the encounter despawned and then the server shut down before it could respawn
+                        me->Respawn();
+                    if (Creature* stalagg = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_STALAGG)))
+                        if (!stalagg->IsAlive())
+                            stalagg->Respawn();
+                    if (Creature* feugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
+                        if (!feugen->IsAlive())
+                            feugen->Respawn();
+                    events.SetPhase(PHASE_NOT_ENGAGED);
+
+                    SetCombatMovement(false);
+
+                    BeginResetEncounter(); // initialize everything properly, and ensure that the coils are loaded by the time we initialize
+                }
```

Comment:
i think its better to call this on reset as its also called on spawn.

leave constructor to initialize variables.